packages/connection-storage/src/connection-provider.ts

```diff
@@ -0,0 +1,80 @@
+import { createContext } from 'react';
+import type { ConnectionInfo } from '@mongodb-js/connection-info';
+import { ConnectionStorage } from './renderer';
+import ConnectionString from 'mongodb-connection-string-url';
+import { merge } from 'lodash';
+
+/**
+ * Exposes an interface that will be implemented for each deployment environment to
+ * get access to the connections on that environment. For Compass Desktop, it will
+ * rely on the ConnectionStorage, for the Data Explorer, on the Atlas API (or another
+ * facade on top of that).
+ */
+export interface ConnectionProvider {
+  listConnections(): Promise<ConnectionInfo[]>;
+  listConnectionHistory?(): Promise<ConnectionInfo[]>;
+  saveConnection?(info: ConnectionInfo): Promise<ConnectionInfo>;
+  deleteConnection?(info: ConnectionInfo): Promise<void>;
+}
+
+export class CompassConnectionProvider implements ConnectionProvider {
+  // Inject the type as it would be an instance (this is literally injecting the singleton)
+  // in case we want to change how ConnectionStorage works in the future.
+  constructor(private readonly storage: typeof ConnectionStorage) {}
+
+  public static defaultInstance(): CompassConnectionProvider {
+    return new CompassConnectionProvider(ConnectionStorage);
+  }
+
+  async listConnections(): Promise<ConnectionInfo[]> {
+    const allConnections = await this.storage.loadAll();
+    return allConnections
+      .filter((connection) => connection.savedConnectionType === 'favorite')
+      .sort(this.sortedAlphabetically);
+  }
+
+  async listConnectionHistory(): Promise<ConnectionInfo[]> {
+    const allConnections = await this.storage.loadAll();
+    return allConnections
+      .filter(
+        (connection) =>
+          connection.savedConnectionType === 'recent' ||
+          !connection.savedConnectionType
+      )
+      .sort(this.sortedAlphabetically);
+  }
+
+  async saveConnection(info: ConnectionInfo): Promise<ConnectionInfo> {
+    const oldConnectionInfo = await this.storage.load({ id: info.id });
+    const infoToSave = oldConnectionInfo
+      ? merge(oldConnectionInfo, info)
+      : info;
+
+    this.ensureWellFormedConnectionString(
+      infoToSave.connectionOptions?.connectionString
+    );
+
+    await this.storage.save({ connectionInfo: infoToSave });
+    return infoToSave;
+  }
+
+  async deleteConnection(info: ConnectionInfo): Promise<void> {
+    await this.storage.delete(info);
+  }
+
+  private ensureWellFormedConnectionString(connectionString: string) {
+    new ConnectionString(connectionString);
+  }
+
+  private sortedAlphabetically = (
+    a: ConnectionInfo,
+    b: ConnectionInfo
+  ): 1 | -1 => {
+    const aName = a.favorite?.name?.toLocaleLowerCase() || '';
+    const bName = b.favorite?.name?.toLocaleLowerCase() || '';
+    return bName < aName ? 1 : -1;
```

Comment:
This was missing before, and probably not a big real-world concern, but this should probably be:

```suggestion
    if (aName === bName) return 0;
    return bName < aName ? 1 : -1;
```

Reply:
Ah, good point. Maybe I can just change it to use localeCompare now that I see it.